ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/database/DatabaseTimerPersistence.java

```diff
@@ -847,37 +823,40 @@ public void run() {
                             while (resultSet.next()) {
+                                String id = null;
                                 try {
-                                    String id = resultSet.getString(1);
+                                    id = resultSet.getString(1);
                                     if (!existing.remove(id)) {
-                                        final Holder holder = timerFromResult(resultSet, timerService);
+                                        final Holder holder = timerFromResult(resultSet, timerService, id, null);
                                         if(holder != null) {
                                             synchronized (DatabaseTimerPersistence.this) {
                                                 knownTimerIds.get(timedObjectId).add(id);
                                                 listener.timerAdded(holder.timer);
                                             }
                                         }
                                     } else {
-                                        final Holder holder = timerFromResult(resultSet, listener.getTimerService());
-                                        if (holder != null) {
-                                            TimerImpl oldTimer = listener.getTimerService().getTimer(id);
-                                            // if it is already in memory but it is not in sync we have a problem
-                                            // remove and add -> the probable cause is db glitch
-                                            EnumSet<TimerState> valid = EnumSet.of(TimerState.IN_TIMEOUT, TimerState.RETRY_TIMEOUT, TimerState.CREATED, TimerState.ACTIVE);
-                                            boolean validDBTimer = valid.contains(holder.timer.getState());
-                                            boolean validMemoryTimer = oldTimer != null && !valid.contains(oldTimer.getState());
-                                            // if timers memory - db are in non intersect subsets of valid/invalid states. we put them in sync
-                                            if (validMemoryTimer && validDBTimer) {
-                                                synchronized (DatabaseTimerPersistence.this) {
-                                                    knownTimerIds.get(timedObjectId).add(holder.timer.getId());
-                                                    listener.timerSync(oldTimer, holder.timer);
+                                        TimerImpl oldTimer = timerService.getTimer(id);
+                                        // if it is already in memory but it is not in sync we have a problem
+                                        // remove and add -> the probable cause is db glitch
+                                        boolean validMemoryTimer = oldTimer != null && !TimerState.CREATED_ACTIVE_IN_TIMEOUT_RETRY_TIMEOUT.contains(oldTimer.getState());
```

Comment:
Is this the right name for this var? It seems more like 'invalidMemoryTimer'. I know it was already the name of the var, so changing it is not necessary to merge this; I'm just pointing it out because if it's not the best name it made the pretty straightforward logic here seem a kind of mysterious.

Reply:
yeah, I felt the same. I've just renamed it to `invalidMemoryTimer` and hope it makes it a bit clearer.